bpe.go

```diff
@@ -0,0 +1,110 @@
+package bpe
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+
+	"github.com/sirupsen/logrus"
+)
+
+type TokenId int32
+
+type DecodedToken []TokenId
+
+type Rule struct {
+	left   TokenId
+	right  TokenId
+	result TokenId
+}
+
+type SpecialTokens struct {
+	unk TokenId
+	pad TokenId
+	bos TokenId
+	eos TokenId
+}
+
+type Model struct {
+	char2id       map[rune]TokenId
+	id2char       map[TokenId]rune
+	rules         []Rule
+	recipe        map[TokenId]DecodedToken
+	revRecipe     map[string]TokenId
+	specialTokens SpecialTokens
+}
+
+func NewModel(nRules int) *Model {
+	return &Model{
+		make(map[rune]TokenId),
+		make(map[TokenId]rune),
+		make([]Rule, nRules),
+		make(map[TokenId]DecodedToken),
+		make(map[string]TokenId),
+		SpecialTokens{-1, -1, -1, -1},
+	}
+}
+
+func DecodedTokenToString(token DecodedToken, id2char map[TokenId]rune) (string, error) {
+	word := ""
+	for _, id := range token {
+		if char, ok := id2char[id]; ok {
+			word = word + string(char)
+		} else {
+			logrus.Fatalf("%d key not found in id2char", id)
+		}
+	}
+	return word, nil
+}
+
+func ReadModel(reader io.Reader) (*Model, error) {
+	scanner := bufio.NewScanner(reader)
```

Comment:
Can you please implement the binary format according to https://github.com/VKCOM/YouTokenToMe/pull/19

You are going to use `encoding/binary.BigEndian`

Reply:
Sure, should I do it here or in a separate PR? Imo it's good to have both loading from text and from binary.

Reply:
There will be no text format at all, so you should rewrite this part.

Reply:
There will be no original, we are hard-forking, and there will be several breaking changes like that. If you want, you can write a converter from text to bin as a separate tool **in the future**.